daemon/graphdriver/windows/windows.go

```diff
@@ -129,61 +226,131 @@ func (d *WindowsGraphDriver) Get(id, mountLabel string) (string, error) {
 		dir = d.dir(id)
 	}
 
-	d.active[id]++
-
 	return dir, nil
 }
 
-func (d *WindowsGraphDriver) Put(id string) error {
+func (d *Driver) Put(id string) error {
 	logrus.Debugf("WindowsGraphDriver Put() id %s", id)
 
+	rId, err := d.resolveId(id)
+	if err != nil {
+		return err
+	}
+
 	d.Lock()
 	defer d.Unlock()
 
-	if d.active[id] > 1 {
-		d.active[id]--
-	} else if d.active[id] == 1 {
-		if err := hcsshim.DeactivateLayer(d.info, id); err != nil {
+	if d.active[rId] > 1 {
+		d.active[rId]--
+	} else if d.active[rId] == 1 {
+		if err := hcsshim.UnprepareLayer(d.info, rId); err != nil {
 			return err
 		}
-		delete(d.active, id)
+		if err := hcsshim.DeactivateLayer(d.info, rId); err != nil {
+			return err
+		}
+		delete(d.active, rId)
 	}
 
 	return nil
 }
 
-func (d *WindowsGraphDriver) Cleanup() error {
+func (d *Driver) Cleanup() error {
 	return nil
 }
 
 // Diff produces an archive of the changes between the specified
 // layer and its parent layer which may be "".
-func (d *WindowsGraphDriver) Diff(id, parent string) (arch archive.Archive, err error) {
-	return nil, fmt.Errorf("The Windows graphdriver does not support Diff()")
+func (d *Driver) Diff(id, parent string) (arch archive.Archive, err error) {
+	rId, err := d.resolveId(id)
+	if err != nil {
+		return
+	}
+
+	// Getting the layer paths must be done outside of the lock.
+	layerChain, err := d.getLayerChain(rId)
+	if err != nil {
+		return
+	}
+
+	d.Lock()
+
+	// To support export, a layer must be activated but not prepared.
+	if d.info.Flavour == filterDriver {
+		if d.active[rId] == 0 {
+			if err = hcsshim.ActivateLayer(d.info, rId); err != nil {
+				d.Unlock()
+				return
+			}
+			defer func() {
+				if err := hcsshim.DeactivateLayer(d.info, rId); err != nil {
+					logrus.Warnf("Failed to Deactivate %s: %s", rId, err)
+				}
+			}()
+		} else {
+			if err = hcsshim.UnprepareLayer(d.info, rId); err != nil {
+				d.Unlock()
+				return
+			}
+			defer func() {
+				if err := hcsshim.PrepareLayer(d.info, rId, layerChain); err != nil {
+					logrus.Warnf("Failed to re-PrepareLayer %s: %s", rId, err)
+				}
+			}()
+		}
+	}
+
+	d.Unlock()
```

Comment:
Yeah, in order to perform the operations to produce the Diff, the layer needs to be activated but not prepared (so mounted but without the kernel storage filter attached).  If the layer has gone through Get, then we need to unprepared it and defer a prepare until after the diff is complete.  If the layer has not gone through Get, then it needs to be activated and a deactivate deferred.